### ISSUE TITLE
Chapter 9 get_tensor_by_name

### ISSUE THREAD

**Issue by nitml** [ageron/handson-ml#261]
i'm not able to understand this line in chapter 9 jupyter notebook in Saving and Restoring Model Section:

```python
theta = tf.get_default_graph().get_tensor_by_name("theta:0") 
```

can you explain it .What is get_tensor_by_name used for ? i tried reading official page but didn't understood completely.I think get_tensor_by_name is used to get  a tensor but then what is ":0" in it ?

**Comment by ageron** [comment]
Hi @nitml,

A TensorFlow graph contains operations and tensors.  The operations represent the computations, and the tensors represent their inputs and outputs.  The name of an operation can be whatever you want, for example:

```python
import tensorflow as tf
a = tf.add(1.0, 2.0, name="addition")
```

When you create an operation, TensorFlow also creates its output tensors and it returns them.  So for example, in the code above, the `addition` operation is created, and its output tensor is returned.  So `a` is the output tensor, not the operation itself.
Tensors have the same name as the operation that outputs them, plus an output index.  The first output tensor has a `:0` suffix.  The second output tensor (for operations that have more than one output) has a `:1` suffix, and so on.

So the code just gets the default graph, and finds the tensor named `theta:0`.  This represents the value (i.e., the "output") of the `theta` variable.

Hope this helps.

**Comment by nitml** [comment]
Great explanation....Thanks for your time .......i'm getting error when i tried to use "ls -l tf_logs/run*" for visualizing Graph and training curve.I used this code in jupyter notebook: 
`reset_graph()
from datetime import datetime
now = datetime.utcnow().strftime("%Y%m%d%H%M%S")
root_logdir = "tf_logs"
logdir = "{}/run-{}/".format(root_logdir, now)
n_epochs = 1000
learning_rate = 0.01

X = tf.placeholder(tf.float32, shape=(None, n + 1), name="X")
y = tf.placeholder(tf.float32, shape=(None, 1), name="y")
theta = tf.Variable(tf.random_uniform([n + 1, 1], -1.0, 1.0, seed=42), name="theta")
y_pred = tf.matmul(X, theta, name="predictions")
with tf.name_scope("loss") as scope:
    error = y_pred - y
    mse = tf.reduce_mean(tf.square(error), name="mse")
optimizer = tf.train.GradientDescentOptimizer(learning_rate=learning_rate)
training_op = optimizer.minimize(mse)

init = tf.global_variables_initializer()
mse_summary=tf.summary.scalar('MSE',mse)
file_writer=tf.summary.FileWriter(logdir,tf.get_default_graph())
n_epochs = 10
batch_size = 100
n_batches = int(np.ceil(m / batch_size))
with tf.Session() as sess:
    sess.run(init)
    for epoch in range(n_epochs):
        for batch_index in range(n_batches):
            X_batch,y_batch=fetch_batch(epoch,batch_index,batch_size)
            if batch_index%10==0:
                summary_str=mse_summary.eval(feed_dict={X:X_batch,y:y_batch})
                step=epoch*n_batches+batch_index
                file_writer.add_summary(summary_str,step)
            sess.run(training_op,feed_dict={X:X_batch,y:y_batch})
    best_theta=theta.eval()         `

it got complied and showed no error  and after that i opened command prompt and typed "ls -l tf_logs/run*"  and i'm getting error even through ml_path is correct in command shell. 
Error is "'   ls' is not recognized as an internal or external command,
operable program or batch file.   " 
Can you please help me ?

**Comment by ageron** [comment]
When you read or write from a file, there is usually a buffer.  When writing to a file, calling the `flush()` method ensures that all the buffer is written to disk immediately (or else this might happen later, for example when the buffer is full, or after some delay).  When you call `close()`, it should call `flush()` before closing the file, so normally there is no reason to call `flush()` just before calling `close()`, it is redundant.  Perhaps I ran into a bug where there was missing data in the events file and I tried several hypotheses, including that the buffer was not flushed properly.  Anyway, I'm not sure why I put a `flush()` there, I think you can safely remove it (but it does not hurt).

Regarding the `as scope`, you can also remove it since I don't seem to use it in the `with` block or after it.  Using the `as` keyword is useful when you need to give an object a name so you can handle it later on. For example:

```python
with open("test", "w") as f:
    f.write("test")
```

Hope this helps,
Aurélien
